### PR TITLE
Faster ParseHeaders

### PR DESCRIPTION
**Pull request by EgorBo**
Judging by the native traces from Platform-Plaintext TE benchmark (linux-x64) it seems we spend some noticeable time in `ParseHeaders`:

### Linux-x64-Xeon:
image

### Linux-arm64-v8.0
image

The current algorithm walks span of data to find "\r\n" and then it tries to extract name and value while doing some validation checks and trimming value. My implementation walks span just once, first it tries to find ":" using a sort of `IndexOfAny(span, ':', ' ', '\t', '\n', '\r')` function to find `':'`'s position and makes sure none of illegal symbols come before it. Then, it tries to find `\r\n` starting from the previously found `:`'s position. After that we try to trim the value from both sides via plain loops. I assume in most cases the value will only have a single space or tab after ":".

I haven't done Arm64 support (waiting for some feedback on this one first) and haven't tested AVX vs SSE yet, but from what I have here it seems to show quite some stable improvements for Platform-Plaintext (up to +300_000 RPS resulting in **12M** requests per second):
```
                                             baseline                                mychanges               diff
| ---------------------- | ------------------------------------ | ------------------------------------ | ------- |
| CPU Usage (%)          |                                   92 |                                   95 |  +3.26% |
| Cores usage (%)        |                                2,578 |                                2,671 |  +3.61% |
| Working Set (MB)       |                                   38 |                                   37 |  -2.63% |
| Private Memory (MB)    |                                  370 |                                  370 |   0.00% |
| Start Time (ms)        |                                    0 |                                    0 |         |
| First Request (ms)     |                                   64 |                                   59 |  -7.81% |
| Requests/sec           |                           11,671,064 |                           11,988,862 |  +2.72% |
| Requests               |                          176,234,240 |                          180,850,384 |  +2.62% |
| Mean latency (ms)      |                                 1.13 |                                 1.14 |  +0.88% |
| Max latency (ms)       |                                57.24 |                                60.75 |  +6.13% |
| Bad responses          |                                    0 |                                    0 |         |
| Socket errors          |                                    0 |                                    0 |         |
| Read throughput (MB/s) |                             1,402.88 |                             1,443.84 |  +2.92% |
| Latency 50th (ms)      |                                 0.71 |                                 0.69 |  -2.66% |
| Latency 75th (ms)      |                                 1.07 |                                 1.04 |  -2.80% |
| Latency 90th (ms)      |                                 1.74 |                                 1.85 |  +6.32% |
| Latency 99th (ms)      |                                14.66 |                                21.68 | +47.89% |
```
_^ peak RPS for my changes was `12,017,325`_

This benchmark is quite stable on Linux-x64 server (but Latency's metrics are not🙂) and the improvements show up every run
image

### Test methodology
I was using crank like this:
```
crank --profile aspnet-citrine-lin \
 --application.framework net7.0 \
 --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/scenarios/platform.benchmarks.yml \
 --scenario plaintext \
 --json t1.json \
 --application.options.outputFiles "/path/to/Microsoft.AspNetCore.Server.Kestrel.Core.dll"
```
where `Microsoft.AspNetCore.Server.Kestrel.Core.dll` is either a "baseline" or "baseline + my changes"

### TODO  
- [ ] Run all [aspnet](https://github.com/aspnet/Benchmarks/tree/main/scenarios) benchmarks, not just Platform-Plaintext
- [ ] Compare AVX vs SSE
- [ ] Add Arm64's AdvSimd (mostly, a copy-paste from SSE)
- [ ] Is `"1:\r\n"`header-value line valid?

Also, I didn't optimize the Multi-span case (where header is split between two reads) but from what I see it happens quite rarely (I can print some statistics from our benchmarks if you need it).

### CODE REVIEW

**Comment by jkotas** [comment]
Related discussions:

- https://github.com/dotnet/aspnetcore/issues/4720
- https://github.com/dotnet/aspnetcore/issues/20518#issuecomment-608930569

**Comment by EgorBo** [comment]
Thanks, didn't see those
Waiting for some thoughts on this before I continue then, I don't mind giving up on this if it's against the strategy of removing Unsafe from Kestrel, it was just a weekend project.

Plaintex-Plaintext benchmark (which is not as heavily optimized as Platform one and closer to the way normal developers write code) got some nice improvements too:
```
```

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs [hunk 1]
```diff
@@ -138,6 +141,221 @@ private void ParseRequestLine(TRequestHandler handler, ReadOnlySpan<byte> reques
...
+            //     return (int)offset;
+            //
+            // (or a switch operator)
+            //
+            // but unfortunately neither JIT nor Roslyn are smart enough to lower it to bit-test
+            // PS: JIT still emits sub-optimal codegen here and doesn't recognize 'bt'
+            //
+            val = (byte)(val - 9); // lower limit is '\t'
+            if (val > 49) // upper limit is ':'
+                return -1;
```
Comment (ltrzesniewski) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs:212]: Shouldn't that be `continue` instead of `return`?

**Comment by benaadams** [comment]
Couldn't you use the bitmask to refer to the second find, rather than searching again? 

i.e. the vector match has found all the matches; so can look at them one after another until run out of matched bits and then vector search again?

**Comment by EgorBo** [comment]
> Couldn't you use the bitmask to refer to the second find, rather than searching again?
> 
> i.e. the vector match has found all the matches; so can look at them one after another until run out of matched bits and then vector search again?

Not sure I understand, do you mean like if we found ":" let's check what else we found in the current vector?
I guess it's unlikely we'll find '\r' in it as headers are usually not that small

**Comment by EgorBo** [comment]
image

This `IndexOf(val)` next to ParseHeaders is a different beast, it is from `SequenceReader::TryReadTo(span, '\n')`. After some additional modifications I managed to get + 150_000 RPS more resulting in 12_173_239 RPS. Just FYI.

I'd love to prototype an optimization for JIT someday where for some high-perf primitives like IndexOf, Memmove, etc we will be inserting  additional polling in tier0, e.g. `int index = ProbeIndexOf(IndexOf(...));` so by the time we promote a method to tier1 we know exactly what kind of values IndexOf usually returns. And if those values are usually big - we might replace `IndexOf` with `IndexOf_HeavilyVectorized` and vice versa: small values - we can even inline a simple loop with a fallback, etc.
It should not be difficult to extend existing Dynamic PGO infrastructure to do it.

**Comment by EgorBo** [comment]
> /benchmark plaintext aspnet-citrine-lin kestrel

Wow! Is that a new thing? Can we have it in dotnet/runtime 😄

**Comment by sebastienros** [comment]
@EgorBo maybe you should pay attention to your emails ...

**Comment by EgorBo** [comment]
> @EgorBo maybe you should pay attention to your emails ...

Oops, I'm sorry, somehow I missed that one 😢

**Comment by sebastienros** [comment]
The benchmark failed because of code design rules. I will need to improve the reports to display the build errors if any instead of this random stack trace.

```
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(155,9): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(181,9): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(185,9): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(188,9): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(210,13): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(215,13): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(233,9): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(247,9): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(251,9): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(254,9): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(266,13): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(293,9): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(297,9): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(303,9): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(309,9): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(319,9): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(323,9): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(327,9): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(335,13): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]
C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Internal\Http\HttpParser.cs(344,13): error IDE0011: Add braces to 'if' statement. [C:\Users\cloudtest\AppData\Local\Temp\aspnetcore\src\Servers\Kestrel\Core\src\Microsoft.AspNetCore.Server.Kestrel.Core.csproj]

```

**Comment by EgorBo** [comment]
> The benchmark failed because of code design rules. I will need to improve the reports to display the build errors if any instead of this random stack trace.

will fix the formatting

**Comment by sebastienros** [comment]
> The user 'EgorBo' is not allowed to perform this action.

;) Does it mean you can't actually merge PRs on this repository?

Also I might need to render this message too.

**Comment by EgorBo** [comment]
hm.. weird, I did see numbers like 12M with this PR

**Comment by halter73** [comment]
@sebastienros Can we run a benchmark with more request headers? Something like the Chrome headers referenced earlier?

```
Host: 127.0.0.1:5001
Connection: keep-alive
Cache-Control: max-age=0
sec-ch-ua: " Not A;Brand";v="99", "Chromium";v="96", "Google Chrome";v="96"
sec-ch-ua-mobile: ?0
sec-ch-ua-platform: "Windows"
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
Sec-Fetch-Site: none
Sec-Fetch-Mode: navigate
Sec-Fetch-User: ?1
Sec-Fetch-Dest: document
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
```

**Comment by sebastienros** [comment]
/benchmark plaintext aspnet-citrine-lin,aspnet-citrine-win,aspnet-citrine-amd,aspnet-perf-lin kestrel

**Comment by EgorBo** [comment]
> <html>
<body>


11,723,936 | 12,032,056 | +2.63%
-- | -- | --



</body>
</html>

much better 😄

**Comment by sebastienros** [comment]
/benchmark plaintext aspnet-citrine-win,aspnet-citrine-amd,aspnet-perf-lin kestrel

**Comment by EgorBo** [comment]
@halter73 so is this PR worth finishing? It almost is, it only reports invalid headers differently so the tests don't like it.

I am asking because of https://github.com/dotnet/aspnetcore/pull/39216#issuecomment-1002285614

**Comment by EgorBo** [comment]
@sebastienros could you please kick an citrine-arm run? I expect to see 2-3% improvement

**Comment by sebastienros** [comment]
Good news is that the benchmark worked, right ;)

**Comment by EgorBo** [comment]
btw how do you deal with flaky benchmarks? I see that PlatformPlaintext jumps from 6.67M RPS to 6.3M on arm - it's more than 5% difference 😮

**Comment by sebastienros** [comment]
The arm machine we use doesn't produce stable results unfortunately. The ones on INTEL are much more stable. I could add more iterations automatically for ARM but I'd rather just wait for the new machines.

**Comment by EgorBo** [comment]
> The arm machine we use doesn't produce stable results unfortunately. The ones on INTEL are much more stable. I could add more iterations automatically for ARM but I'd rather just wait for the new machines.

Yeah if I run it multiple times I see almost 6.8M RPS locally (with a fix for faster comparison against zero vector) while the best official result is ~6.4M

**Comment by halter73** [comment]
@davidfowl Assuming we get a consistent >2% improvement in the modified plaintext benchmark (using Chrome request headers), would we be willing to take the unsafe code? I would be personally.

If we take it (no promises yet), we'll need to make sure we don't break on big-endian architectures. We should also reject the same requests we did before (e.g. `"MyHeader:\r\n"` should continue to fail even though `"MyHeader: \r\n"` doesn't). Don't feel compelled to do this before getting solid confirmation we're willing to take new unsafe code.

**Comment by japui** [comment]
ParseHeaders mais rápidos

**Pull request closed without merging**